DOC: Complete API for AMI analyze and normalize

pllim opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Towards [JP-4107](https://jira.stsci.edu/browse/JP-4107)

<!-- If this PR closes a GitHub issue that is not attached to a Jira ticket, uncomment the line below, and reference it here by its number -->
<!-- Closes # -->

<!-- describe the changes comprising this PR here -->
This PR addresses AMI portion of https://github.com/spacetelescope/jwst/issues/9793 . Only touch docs so should not need RT.

On main:

* https://jwst.readthedocs.io/en/latest/jwst/ami_analyze/index.html
* https://jwst.readthedocs.io/en/latest/jwst/ami_normalize/index.html

With this patch:

* https://jwst-pipeline--10095.org.readthedocs.build/en/10095/jwst/ami_analyze/index.html
* https://jwst-pipeline--10095.org.readthedocs.build/en/10095/jwst/ami_normalize/index.html

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [ ] update or add relevant tests
  - [x] update relevant docstrings and / or `docs/` page
  - [ ] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

pllim commented on jwst/ami/leastsqnrm.py:
```diff
@@ -323,14 +327,14 @@ def populate_antisymmphasearray(deltaps, n=7):
-    fringephasearray[3,q+4:] = coeffs[15:18]
-    fringephasearray[4,q+5:] = coeffs[18:20]
-    fringephasearray[5,q+6:] = coeffs[20:]
+        fringephasearray[0, q + 1 :] = coeffs[0:6]
```
I don't remember adding spaces around `q + N`. The pre-commit must have done it...

pllim reviewed (commented):
Whatever layout and presentation style we decide on in this PR would be used as a basis for the rest of the package. Doing the entire package at once would make the PR hard to review and also drag on for too long, causing conflict and wasting developer time, so let's start with AMI (alphabetically selected).

My main priority is fixing doc build warning. Any violation that does not trigger warning (e.g., malformed listing), I will fix if I happen to notice or be asked to only. Beautifying and vetting API doc contents is out of scope here (and would also drag things on). My ticket is only to expose them.

For my sanity, I also alphabetized the intersphinx listing in this PR.

emolter reviewed (commented):
Overall I'm on board with exposing all our API, but I feel pretty strongly that we need to define a deprecation policy, i.e write down what we consider "public" in the sense that we won't change it without warning.  It's one thing to change API that's not documented in a minor release with no change log, which we've been doing for a long time, and another to change that API when it's documented in the readthedocs.

edit: lol, went to my next notification and it's a PR from you that would do the above.  I guess this is just a note then that I'd probably prefer if that one went in first.

I've seen a few typos and things that don't render right within docs that you didn't touch here.  Do you want me to flag those, and basically have a careful read through everything that's newly rendered here?  If so, it'll take longer - there are a lot of new docstrings exposed here.

pllim commented on docs/jwst/ami_analyze/api_ref.rst:
```diff
@@ -0,0 +1,55 @@
+API
+===
+
+Main API
```
I didn't really that the existing behavior of dumping API directly on the main page. It doesn't look pretty.

But I agree "Main API" and "Other API" also aren't very nice. Would like to hear if others have a preference. And when we reach consensus, I can change it.

melanieclarke replied:
> something like "Public Step API" and "Complete Developer API"

I like this idea.

pllim replied:
https://jwst.readthedocs.io/en/latest/jwst/user_documentation/more_information.html#api-public-vs-private is live. Re-titled as suggested. Thanks!

emolter commented:
> @melanieclarke told me to prioritize this ticket first. But that said, the whole ticket is going to take a while. This is just AMI stuff. So chances are the policy (https://github.com/spacetelescope/jwst/pull/10096) would actually go in first before this ticket is done as a whole. Not sure if this elevates your concern or not.

I guess the expectation is that both will get done within the build, so it doesn't really matter.  I want to avoid though a situation where there's a long period where all this API is exposed but we haven't told users anywhere that it doesn't follow standard deprecation rules.

emolter commented:
The one clean-up thing I happened to notice that's a bit egregious is for the one-liner descriptions of `jwst.ami.analyticnrm2.asffringe` and `jwst.ami.analyticnrm2.asf_hex`, it gets truncated after the word A.K.A., presumably because it considers that punctuation.  I'd suggest changing those to AKA so that the rest of the one-liner gets rendered.

pllim commented:
Re: https://github.com/spacetelescope/jwst/pull/10095#issuecomment-3675419661

I removed the offending periods and added Ned as commit co-author. I will mark the comment as resolved to ease future review handling. Thanks!

emolter commented on docs/jwst/ami_normalize/api_ref.rst:
```diff
@@ -0,0 +1,17 @@
+Complete Developer API
+======================
+
+(Also see :ref:`Other API in ami_analyze <ami_analyze_other_api>`.)
```
This is no longer called Other API
```suggestion
The developer API is shared with the ``ami_normalize`` step.  See :ref:`AMI Normalize developer API <ami_analyze_other_api>` for more information.
```

emolter commented on jwst/ami/instrument_data.py:
```diff
@@ -109,14 +115,14 @@ def read_data_model(self, input_model):
         Parameters
         ----------
-        input_model : instance Data Model
+        input_model : `~stdatamodels.DataModel`
```
Did we decide yet whether these should all point to the datamodel base class or to JwstDataModel?

emolter replied:
I agree it's out of scope, but it might save someone some time if we already knew.  I'm happy with it as-is

pllim replied:
I will just change this one to JwstDataModel now while I am at it. Do you want a new GH issue to discuss the package wide question?

emolter replied:
That would not hurt, but IMO it's not very high priority either

emolter commented on jwst/ami/instrument_data.py:
```diff
@@ -109,14 +115,14 @@ def read_data_model(self, input_model):
         scidata_ctrd : numpy array
             Cropped, centered, optionally cleaned AMI data
-        dqmask_ctrd :
+        dqmask_ctrd
```
what's the reason for removing the `:` here?

emolter commented on jwst/ami/lg_model.py:
```diff
@@ -158,23 +162,27 @@ def make_model(self, fov, psf_offset=(0, 0)):
-        psf_offset : detector pixels
-            Center offset from center of array
+        psf_offset : tuple of int
+            Detector pixels defining the center offset from center of array
```
```suggestion
            Offset from center of array in units of detector pixels.
```

emolter commented on jwst/ami/lg_model.py:
```diff
@@ -107,24 +110,25 @@ def __init__(
-        psf_offset : detector pixels
-            Center offset from center of array
+        psf_offset : tuple of int
+            Detector pixels defining the center offset from center of array
```
```suggestion
            Offset from center of array in units of detector pixels.
```

pllim replied:
Hmm GH won't let me apply this directly. I can do it manually later.

emolter commented on jwst/ami/matrix_dft.py:
```diff
@@ -19,13 +21,14 @@
+* 2012-05-18: module renamed SFT.py -> matrixDFT.py
+* 2012-09-26: minor big fixes
+* 2015-01-21: Eliminate redundant code paths, correct parity flip,
+  PEP8 formatting pass (except variable names)-- J. Long
```
This code history is incomplete - I know I made substantial changes to this file last year.  Should we remove it entirely, and just credit the various authors and references?  Or make the additional comment that changes post-201X are logged on GitHub?

pllim commented on jwst/ami/matrix_dft.py:
```diff
@@ -18,23 +20,22 @@
-PEP8 formatting pass (except variable names)-- J. Long
+Notes
+-----
+Code originally by A. Sivaramakrishnan, with early modifications by
```
@emolter and/or @melanieclarke , does this look good to you now?

emolter replied:
I'd simplify this further and just attribute original code authorship to A. Sivaramakrishnan, M. Perrin, and J. Long, and then reference Soummer paper separately.

Right now the wording is unclear - "in accordance to" is an unusual construction, and the fact that J. Long comes after the reference makes it seem like J. Long is another reference, not another code author

pllim replied:
Are you able to provide desired text as suggestion? Thanks!

pllim replied:
Nvm I did a best effort interpretation. Hopefully good enough.

pllim commented on jwst/ami/utils.py:
```diff
@@ -49,24 +49,32 @@ class Affine2d:
-    g(x,y) = f(x',y'), where::
+    g(x,y) = f(x',y'), where:
+
+    .. code-block:: text
```
Note for future person: The code block type has to be defined explicitly because package default is set to Python, so this text was getting highlighted in a way that does not make sense.

pllim commented:
Okay, I think I have addressed the comments from yesterday. Please re-review. Thanks!

melanieclarke commented on jwst/ami/matrix_dft.py:
```diff
@@ -18,23 +20,22 @@
+-----
+Code originally by A. Sivaramakrishnan, with early modifications by
+M. Perrin in accordance to Soummer et al. 2007 [1]_ and J. Long.
+Changes since then are kept in Git commit history.
```
I think this is implied, no need to state it.  Otherwise, this change looks good to me.

```suggestion
```

emolter reviewed (approved):
I think all my questions have been answered
